src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -256,90 +269,104 @@ private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
         if (superClassName != null) {
-            final ClassDesc nearest =
-                    getNearestClassWithSameName(superClassName, qualifiedClassName);
-            if (nearest == null) {
-                Optional.ofNullable(innerClasses.get(superClassName))
-                        .ifPresent(ClassDesc::registerNestedSubclass);
-            }
-            else {
-                nearest.registerNestedSubclass();
-            }
+            getNearestClassWithSameName(superClassName, classDesc -> {
+                return CheckUtil.typeDeclarationNameMatchingCount(qualifiedClassName,
+                                                                  classDesc.getQualifiedName());
+            }).ifPresentOrElse(ClassDesc::registerNestedSubclass, () -> {
+                Optional
+                    .ofNullable(innerClasses.get(superClassName))
+                    .ifPresent(ClassDesc::registerNestedSubclass);
+            });
         }
     }
 
     /**
-     * Checks if there is a class with same name.
+     * Register to the super class of anonymous inner class that the given class is instantiated
+     * by an anonymous inner class.
      *
-     * @param className name of the class
-     * @param superClassName name of the super class
-     * @return true if there is another class with same name.
-     * @noinspection CallToStringConcatCanBeReplacedByOperator
+     * @param literalNewAst ast node of {@link TokenTypes#LITERAL_NEW} representing anonymous inner
+     *                      class
+     * @param outerTypeDeclName Fully qualified name of the outer type declaration of anonymous
+     *                          inner class
      */
-    private ClassDesc getNearestClassWithSameName(String className, String superClassName) {
-        final String dotAndClassName = PACKAGE_SEPARATOR.concat(className);
-        return innerClasses.entrySet().stream()
-                .filter(entry -> entry.getKey().endsWith(dotAndClassName))
-                .map(Map.Entry::getValue)
-                .min((first, second) -> {
-                    return getClassDeclarationNameMatchingCountDiff(superClassName, first, second);
-                })
-                .orElse(null);
+    private void registerAnonymousInnerClassToSuperClass(DetailAST literalNewAst,
+                                                         String outerTypeDeclName) {
+        final String superClassName = CheckUtil.getShortNameOfAnonInnerClass(literalNewAst);
+
+        getNearestClassWithSameName(superClassName, classDesc -> {
+            return getAnonSuperTypeMatchingCount(outerTypeDeclName, classDesc.getQualifiedName());
+        }).ifPresentOrElse(ClassDesc::registerSuperClassOfAnonymousInnerClass, () -> {
+            Optional
+                .ofNullable(innerClasses.get(superClassName))
+                .ifPresent(ClassDesc::registerSuperClassOfAnonymousInnerClass);
+        });
```

Comment:
it is better now, but can be improved, like
```
Function getAnonCount = classDesc -> getAnonSuperTypeMatchingCount(outerTypeDeclName, classDesc.getQualifiedName());
getNearestClassWithSameName(superClassName,getAnonCount)
.or(Optional.ofNullable(innerClasses.get(superClassName))
.ifPresent(...)
```

Reply:
Done